Flaky test: test_load_vocabs

http://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-cpu-unittest/detail/PR-692/7/pipeline/81/ 

```
=================================== FAILURES ===================================

_______________________________ test_load_vocabs _______________________________

[gw3] linux -- Python 3.6.7 /var/lib/jenkins/workspace/gluon-nlp-cpu-py3@4/conda/cpu/py3/bin/python



    @pytest.mark.remote_required

    def test_load_vocabs():

>       dataset = SQuAD(segment='dev', root='tests/data/squad')



scripts/tests/test_question_answering.py:70: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

src/gluonnlp/data/question_answering.py:136: in __init__

    super(SQuAD, self).__init__(self._read_data())

src/gluonnlp/data/question_answering.py:174: in _read_data

    samples = json.load(f)

conda/cpu/py3/lib/python3.6/json/__init__.py:299: in load

    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)

conda/cpu/py3/lib/python3.6/json/__init__.py:354: in loads

    return _default_decoder.decode(s)

conda/cpu/py3/lib/python3.6/json/decoder.py:339: in decode

    obj, end = self.raw_decode(s, idx=_w(s, 0).end())

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 



self = <json.decoder.JSONDecoder object at 0x7f64a4946208>

s = '{"data": [{"title": "Super_Bowl_50", "paragraphs": [{"context": "Super Bowl 50 was an American football game to deter...rkers in the poor countries"}, {"answer_start": 239, "text": "low-skilled workers in the poor countries"}, {"answer_st'

idx = 0



    def raw_decode(self, s, idx=0):

        """Decode a JSON document from ``s`` (a ``str`` beginning with

        a JSON document) and return a 2-tuple of the Python

        representation and the index in ``s`` where the document ended.

    

        This can be used to decode a JSON document from a string that may

        have extraneous data at the end.

    

        """

        try:

>           obj, end = self.scan_once(s, idx)

E           json.decoder.JSONDecodeError: Unterminated string starting at: line 1 column 3342327 (char 3342326)



conda/cpu/py3/lib/python3.6/json/decoder.py:355: JSONDecodeError


```

--- a/Makefile
+++ b/Makefile
@@ -19,7 +19,7 @@ ROOTDIR = $(CURDIR)
 MD2IPYNB = $(ROOTDIR)/docs/md2ipynb.py
 
 flake8:
-	flake8 --exclude conda,test_*.py,scripts/word_embeddings/tools/extern --count --select=E901,E999,F821,F822,F823 --show-source --statistics $(lintdir)
+	flake8 --exclude conda,*tests*,test_*.py,scripts/word_embeddings/tools/extern --count --select=E901,E999,F821,F822,F823 --show-source --statistics $(lintdir)
 
 pylint:
 	pylint --rcfile=$(ROOTDIR)/.pylintrc $(lintdir)

--- a/ci/prepare_clean_env.sh
+++ b/ci/prepare_clean_env.sh
@@ -4,6 +4,7 @@ env_name=$1
 export LD_LIBRARY_PATH=/usr/local/cuda/lib64
 export CUDA_VISIBLE_DEVICES=$EXECUTOR_NUMBER
 export CONDA_ENVS_PATH=$PWD/conda
+export CONDA_PKGS_DIRS=$PWD/conda/pkgs
 
 make clean
 conda env update --prune -p conda/${env_name} -f env/${env_name}.yml

--- a/scripts/tests/conftest.py
+++ b/scripts/tests/conftest.py
@@ -1,0 +1,37 @@
+# coding: utf-8
+
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""conftest.py contains configuration for pytest."""
+
+import os
+
+import pytest
+
+import gluonnlp as nlp
+from ..question_answering.data_processing import VocabProvider
+
+
+###############################################################################
+# Datasets
+###############################################################################
+@pytest.fixture(scope='session')
+def squad_dev_and_vocab_provider():
+    path = os.path.join('tests', 'data', 'squad')
+    dataset = nlp.data.SQuAD(segment='dev', version='1.1', root=path)
+    vocab_provider = VocabProvider(dataset)
+    return dataset, vocab_provider

--- a/scripts/tests/test_question_answering.py
+++ b/scripts/tests/test_question_answering.py
@@ -21,7 +21,6 @@ import pytest
 
 from mxnet.gluon.data import DataLoader, SimpleDataset
 
-from gluonnlp.data import SQuAD
 from ..question_answering.data_processing import SQuADTransform, VocabProvider
 
 question_max_length = 30
@@ -29,9 +28,8 @@ context_max_length = 256
 
 
 @pytest.mark.remote_required
-def test_transform_to_nd_array():
-    dataset = SQuAD(segment='dev', version='1.1', root='tests/data/squad')
-    vocab_provider = VocabProvider(dataset)
+def test_transform_to_nd_array(squad_dev_and_vocab_provider):
+    dataset, vocab_provider = squad_dev_and_vocab_provider
     transformer = SQuADTransform(
         vocab_provider, question_max_length, context_max_length)
     record = dataset[0]
@@ -42,9 +40,8 @@ def test_transform_to_nd_array():
 
 
 @pytest.mark.remote_required
-def test_data_loader_able_to_read():
-    dataset = SQuAD(segment='dev', root='tests/data/squad')
-    vocab_provider = VocabProvider(dataset)
+def test_data_loader_able_to_read(squad_dev_and_vocab_provider):
+    dataset, vocab_provider = squad_dev_and_vocab_provider
     transformer = SQuADTransform(
         vocab_provider, question_max_length, context_max_length)
     record = dataset[0]
@@ -66,9 +63,8 @@ def test_data_loader_able_to_read():
 
 
 @pytest.mark.remote_required
-def test_load_vocabs():
-    dataset = SQuAD(segment='dev', root='tests/data/squad')
-    vocab_provider = VocabProvider(dataset)
+def test_load_vocabs(squad_dev_and_vocab_provider):
+    dataset, vocab_provider = squad_dev_and_vocab_provider
 
     assert vocab_provider.get_word_level_vocab() is not None
     assert vocab_provider.get_char_level_vocab() is not None

